lib/types/composite_type.js

```diff
@@ -0,0 +1,129 @@
+'use strict';
+var util = require('util'),
+    DescribedType = require('./described_type'),
+    ForcedType = require('./forced_type');
+
+var knownTypes = {};
+function convertKnownType(described) {
+  if (described === undefined || described === null ||
+      (!(described instanceof DescribedType))) {
+    return described;
+  }
+
+  if (knownTypes.hasOwnProperty(described.descriptor))
+    return new knownTypes[described.descriptor](described);
+  return undefined;
+}
+
+function wrapField(field, value) {
+  if (value instanceof ForcedType) return value;
+  if (value === undefined || value === null) {
+    if (field.mandatory) throw new Error('missing mandatory field: ' + field.name);
+    return null;
+  }
+
+  if (field.type === '*') {
+    if (field.hasOwnProperty('requires'))
+       return (value instanceof field.requires) ? value : new field.requires(value);
+    return value;
+  }
+
+  if (Array.isArray(value)) {
+    if (field.multiple === false) {
+      throw Error('field ' + field.name + ' does not support multiple values, got ' + JSON.stringify(value));
+    }
+
+    if (value.length === 0) return new ForcedType(field.type, null);
+    else if (value.length === 1) return new ForcedType(field.type, value[0]);
+    else {
+      var values = [], _len = value.length;
+      for (var i = 0; i < _len; ++i) {
+        values.push(new ForcedType(field.type, value[i]));
```

Comment:
should this be `values.push(this.wrapField(field, value[i]));`, in case we have e.g. multiple of type '*'?